# Homework Session-09

In this homework, we'll deploy the dogs vs cats model we trained in the previous homework.

Download the model from here:

[alexeygrigorev/large-datasets@dogs-cats-model dogs_cats_10_0.687.h5 (download)](https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats-model/dogs_cats_10_0.687.h5)

In [1]:
import tensorflow as tf
from tensorflow import keras

In [2]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats-model/dogs_cats_10_0.687.h5 -O dogs-cats-model.h5

--2021-12-01 19:54:26--  https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats-model/dogs_cats_10_0.687.h5
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/426348925/f4f8406c-b1cd-4377-94a0-8a89b6f826bd?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211201%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211201T185426Z&X-Amz-Expires=300&X-Amz-Signature=ee8504ffc96656696420bdfa145dc7a83b65f26e0fb49aa9961ef2354d076c0c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=426348925&response-content-disposition=attachment%3B%20filename%3Ddogs_cats_10_0.687.h5&response-content-type=application%2Foctet-stream [following]
--2021-12-01 19:54:26--  https://github-releases.githubusercontent.com/426348925/f4f8406c-b1cd-4377-94a0-8a89b6f826bd?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-A

In [4]:
model = keras.models.load_model('dogs-cats-model.h5')

converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

with open('dogs-cats-model.tflite','wb') as f_out:
    f_out.write(tflite_model)

2021-12-01 20:03:53.855038: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2021-12-01 20:03:53.855084: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
2021-12-01 20:03:53.855102: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (ubuntu): /proc/driver/nvidia/version does not exist
2021-12-01 20:03:53.878102: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-12-01 20:03:55.505418: W tensorflow/python/util/util.cc:348] Sets are not currently considere

INFO:tensorflow:Assets written to: /tmp/tmpz6syd2x3/assets


2021-12-01 20:03:56.509553: I tensorflow/core/grappler/devices.cc:66] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0
2021-12-01 20:03:56.509690: I tensorflow/core/grappler/clusters/single_machine.cc:357] Starting new session
2021-12-01 20:03:56.566377: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:1137] Optimization results for grappler item: graph_to_optimize
  function_optimizer: function_optimizer did nothing. time = 0.753ms.
  function_optimizer: function_optimizer did nothing. time = 0.002ms.

2021-12-01 20:03:57.490047: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:351] Ignored output_format.
2021-12-01 20:03:57.490115: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:354] Ignored drop_control_dependency.
2021-12-01 20:03:57.750643: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:210] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.


## Question 1
Now convert this model from Keras to TF-Lite format.

__What's the size of the converted model?__

In [7]:
!wc -c dogs-cats-model.tflite

44865952 dogs-cats-model.tflite


__*Answer Q1: 44865952*__

## Question 2
To be able to use this model, we need to know the index of the input and the index of the output.

__What's the output index for this model?__

In [10]:
import tensorflow.lite as tflite
interpreter = tflite.Interpreter(model_path='dogs-cats-model.tflite')
interpreter.allocate_tensors()
input_index = interpreter.get_input_details()[0]['index']
print('input_index: %s' % input_index)
output_index = interpreter.get_output_details()[0]['index']
print('output_index: %s' % output_index)

input_index: 0
output_index: 13


__*Answer Q2: 13*__

## Preparing the image
You'll need some code for downloading and resizing images. You can use this code:

```Python
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img
```   
    
For that, you'll need to have pillow installed:

```Python
pip install pillow
```

Let's download and resize this image:

[upload.wikimedia.org/wikipedia/commons/9/9a/Pug_600.jpg](https://upload.wikimedia.org/wikipedia/commons/9/9a/Pug_600.jpg)

__Based on the solution of the previous homework, what should be the target size for the image?__

In [12]:
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

In [13]:
url = 'https://upload.wikimedia.org/wikipedia/commons/9/9a/Pug_600.jpg'
size = (150,150)

img = download_image(url)
img = prepare_image(img, size)

## Question 3
Now we need to turn the image into an numpy array and pre-process it.

Tip: Check the previous homework. What was the pre-processing we did there?

__After the pre-processing, what's the value in the first pixel, the R channel?__

In [41]:
import numpy as np

In [42]:
x = np.array(img)
x = x/255
X = np.array([x])

X

array([[[[0.70588235, 0.77254902, 0.74901961],
         [0.69411765, 0.76078431, 0.72941176],
         [0.63137255, 0.69803922, 0.66666667],
         ...,
         [0.76470588, 0.85098039, 0.83921569],
         [0.68235294, 0.74509804, 0.74509804],
         [0.6745098 , 0.76862745, 0.77647059]],

        [[0.5254902 , 0.59215686, 0.56862745],
         [0.7372549 , 0.80392157, 0.78039216],
         [0.60784314, 0.6745098 , 0.64313725],
         ...,
         [0.79215686, 0.86666667, 0.85882353],
         [0.63137255, 0.70196078, 0.63921569],
         [0.78823529, 0.87843137, 0.85490196]],

        [[0.68627451, 0.74901961, 0.7372549 ],
         [0.61960784, 0.68627451, 0.6627451 ],
         [0.64705882, 0.71372549, 0.68235294],
         ...,
         [0.6       , 0.65882353, 0.64705882],
         [0.62745098, 0.69411765, 0.65490196],
         [0.70588235, 0.79607843, 0.77254902]],

        ...,

        [[0.58039216, 0.61960784, 0.58431373],
         [0.49019608, 0.5254902 , 0.50588235]

__*Answer Q3: 0.70588235*__

## Question 4
__Now let's apply this model to this image. What's the output of the model?__

In [32]:
interpreter.set_tensor(input_index, X)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)
preds

array([[0.7704913]], dtype=float32)

__*Answer Q4: 0.7704913*__

## Prepepare the lambda code
Now you need to copy all the code into a separate python file. You will need to use this file for the next two questions.

Tip: you can test this file locally with ipython or Jupyter Notebook by importing the file and invoking the function from this file.

## Docker
For the next two questions, we'll use a Docker image that I already prepared. This is the Dockerfile that I used for creating the image:

```Python
FROM public.ecr.aws/lambda/python:3.8
COPY cats-dogs-v2.tflite .
And pushed it to agrigorev/zoomcamp-cats-dogs-lambda:v2.
```

Note: The image already contains a model and it's not the same model as the one we used for questions 1-4.

## Question 5
Now let's extend this docker image, install all the required libraries and add the code for lambda.

You don't need to include the model in the image. It's already included. The name of the file with the model is cats-dogs-v2.tflite and it's in the current workdir in the image (see the Dockerfile above for the reference).

__What's the image id of the base image?__

In the build logs (on Linux), you'll see a log like that:

```sh
$ docker some-command-for-building
Sending build context to Docker daemon  2.048kB
Step 1/N : FROM agrigorev/zoomcamp-cats-dogs-lambda:v2
 ---> XXXXXXXXXXXX
Step 2/N : ....
```

You need to get this XXXXXXXXXXXX.

On MacOS and Windows, the logs for docker build are different. To get the image id there, you can use docker image ls -a.

__*Answer Q5: 322fc756f258*__

## Question 6
Now run the container locally.

Score this image: [upload.wikimedia.org/wikipedia/commons/1/18/Vombatus_ursinus_-Maria_Island_National_Park.jpg](https://upload.wikimedia.org/wikipedia/commons/1/18/Vombatus_ursinus_-Maria_Island_National_Park.jpg)

__What's the output from the model?__

__*Answer Q6: 0.5413472652435303*__